pkg/server/api/entry/v1/service.go

```diff
@@ -322,6 +332,158 @@ func (s *Service) GetAuthorizedEntries(ctx context.Context, req *entryv1.GetAuth
+			Entries: entries,
+		})
+	}
+
+	// Prepulate the entry page used in the response with empty entry structs.
+	// These will be reused for each sparse entry response.
+	entryRevisions := make([]*entryv1.EntryRevision, entryPageSize)
+	for i := range entryRevisions {
+		entryRevisions[i] = &entryv1.EntryRevision{}
+	}
+	for i := 0; i < len(entries); {
+		more := false
+		n := len(entries) - i
+		if n > entryPageSize {
+			n = entryPageSize
+			more = true
+		}
+		for j, entry := range entries[i : i+n] {
+			entryRevisions[j].Id = entry.Id
+			entryRevisions[j].RevisionNumber = entry.RevisionNumber
+		}
+
+		if err := stream.Send(&entryv1.SyncAuthorizedEntriesResponse{
+			EntryRevisions: entryRevisions[:n],
+			More:           more,
+		}); err != nil {
+			return err
+		}
+		i += n
+	}
+
+	// Now wait for the client to request IDs that they need the full copy of.
+	// Each request is treated independently. Entries are paged back fully
+	// before the next request is received, using the More field as a flag to
+	// signal to the caller when all requested entries have been streamed back.
+	resp := &entryv1.SyncAuthorizedEntriesResponse{}
+	entriesSorted := false
+	for {
+		req, err := stream.Recv()
+		if err != nil {
+			// EOF is normal and happens when the server processes the
+			// CloseSend sent by the client. If the client closes the stream
+			// before then then Canceled is expected. Either way, these
+			// conditions are normal and not an error.
+			if errors.Is(err, io.EOF) || status.Code(err) == codes.Canceled {
+				return nil
+			}
+			return err
+		}
+
+		if !entriesSorted {
+			// Sort the entries by ID for efficient lookups. This is done
+			// lazily since we only need these lookups if full copies are
+			// being requested.
+			sortEntriesByID(entries)
+			entriesSorted = true
+		}
+
+		// Sort the requested IDs for efficient lookups into the sorted entry
+		// list.
+		sort.Strings(req.Ids)
+
+		// Page back the requested entries. The slice for the entries in the response
+		// is reused to reduce memory pressure. Since both the entries and
+		// requested IDs are sorted, we can reduce the amount of entries we
+		// need to search as we iteratively move through the requested IDs.
+		resp.Entries = resp.Entries[:0]
+		entriesToSearch := entries
+		for _, id := range req.Ids {
+			i, found := sort.Find(len(entriesToSearch), func(i int) int {
+				return strings.Compare(id, entriesToSearch[i].Id)
+			})
+			if found {
+				if len(resp.Entries) == entryPageSize {
+					// Adding the entry just found will exceed our page size.
+					// Ship the pageful of entries first and signal that there
+					// is more to follow.
+					resp.More = true
+					if err := stream.Send(resp); err != nil {
+						return err
+					}
+					resp.Entries = resp.Entries[:0]
+				}
+				resp.Entries = append(resp.Entries, entriesToSearch[i])
+			}
+			entriesToSearch = entriesToSearch[i:]
+			if len(entriesToSearch) == 0 {
```

Comment:
what happens if there are some IDs left, to be returned, but we get out of entries? may we log that? 
and in case entry not found, nothing is either, and slice is "recalculated" with `[0:]` that is doing nothing, but may we avoid that?

Reply:
IMO, the re-slicing to [0:] in this corner case seems ok since it is a valid operation and I think any code added to try and prevent it happening hurts readability....

I'm not sure about logging either. This endpoint is intended for agents and a well-behaving agent won't make a request with IDs that it didn't observe in the initial response. The code is robust against that happening but in practice it would be strange for that to happen.... how strongly do you feel about the logging? I am happy to add it if you are very worried.